cluster_generator/data_structures.py

```diff
@@ -1,100 +1,560 @@
+    "dm_velocity_x": ("dark_mater_density", 0, "velocity"),
+    "dm_velocity_y": ("dark_mater_density", 1, "velocity"),
+    "dm_velocity_z": ("dark_mater_density", 2, "velocity"),
+    "mixing": ("density", None, "model_id"),
+}
+
+#: Field standard units.
+units = {
+    "density": ("Msun/kpc**3", None),
+    "pressure": ("Msun/kpc/Myr**2", None),
+    "dark_matter_density": ("Msun/kpc**3", None),
+    "stellar_density": ("Msun/kpc**3", None),
+    "temperature": ("K", "thermal"),
+    "gravitational_potential": ("kpc**2/Myr**2", None),
+    "velocity_x": ("kpc/Myr", None),
+    "velocity_y": ("kpc/Myr", None),
+    "velocity_z": ("kpc/Myr", None),
+    "magnetic_field_strength": ("G", None),
+    "stellar_velocity_x": ("kpc/Myr", None),
+    "stellar_velocity_y": ("kpc/Myr", None),
+    "stellar_velocity_z": ("kpc/Myr", None),
+    "dm_velocity_x": ("kpc/Myr", None),
+    "dm_velocity_y": ("kpc/Myr", None),
+    "dm_velocity_z": ("kpc/Myr", None),
+    "mixing": ("", None),
+}
+_allowed_fields = (
+    additive_fields + dens_weighted_fields + list(dens_weighted_nonfields.keys())
+)  # fields that are recognized.
+
+
+class YTHDF5:
+    """
+    Wrapper class for YT style HDF5 files. Used to manage the writing of :py:class:`model.ClusterModel` instances
+    to YT datasets.
+    """
+
+    def __init__(
+        self, domain_dimensions, bbox, chunksize=64, fields=None, filename=None
+    ):
+        """
+        Initializes the :py:class:`YTHDF5` instance.
+
+        Parameters
+        ----------
+        domain_dimensions: array-like
+            The dimensions of the grid along each axis (3-tuple)
+        bbox: array-like
+            The bounding box of the grid. Should be of size ``(3,2)``.
+        chunksize: int, optional
+            The maximum size of a chunk. Chunking is used to conserve memory during computations. A higher ``chunksize`` will
+            increase the memory usage but decrease computation time, a lower value will do the opposite. Default is ``64``.
+        fields: list of str, optional
+            The fields to include in the dataset. If ``None``, then all of the available fields are included.
+        filename: str, optional
+            The filename for the underlying file. If left as ``None``, a ``tmpfile`` is generated to hold the dataset.
+            Otherwise, if a ``str`` instance is provided, then that file will be used. If a pre-existing file is specified,
+            then it will either overwrite or raise an error depending on ``overwrite``.
+        """
+        self._cm = None
+        self._dd = np.array(domain_dimensions, dtype="uint32")
+        self.bbox = np.array(bbox, dtype="float64")
+        self.chunksize = chunksize
+        self.fields = fields if fields is not None else _allowed_fields
+        self._buffer = None
+        self._gridbuffer = None
+
+        if filename is None:
+            tf = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
Shouldn't we always require a file to be specified? Or is there a case where we would want a temp file?